provisioner/aws-state.js

```diff
@@ -0,0 +1,292 @@
+}
+
+
+/**
+ * Get the raw state
+ */
+AwsState.prototype.get = function(region, type) {
+  if (region && type) {
+    return this.__state[region][type];
+  } else if (region && !type) {
+    return this.__state[region];
+  } else if (!region && !type) {
+    return this.__state;
+  }
+};
+
+
+
+/**
+ * Return a list of workerTypes known to AWS
+ */
+AwsState.prototype.knownWorkerTypes = function() {
+  var workerTypes = [];
+  var that = this;
+
+  this.regions().forEach(function(region) {
+    that.typesForRegion(region).forEach(function(workerType) {
+      if (workerTypes.indexOf(workerType) === -1) {
+        workerTypes.push(workerType);
+      }
+    });
+  });
+
+  return workerTypes;
+
+};
+
+
+/**
+ * Return a list of all running Instance Ids that are known in this AWS State
+ * These are not categorized by region.  It's one list of strings.
+ */
+AwsState.prototype.listRunningInstanceIds = function() {
+  var allIds = [];
+  var that = this;
+
+  this.regions().forEach(function(region) {
+    that.typesForRegion(region).forEach(function(workerType) {
+      var ids = that.__state[region][workerType].running.map(function(x) {
+        return x.InstanceId;
+      });
+      Array.prototype.push.apply(allIds, ids);
+    });
+  });
+
+  return allIds;
+};
+
+
+/**
+ * Return a list of all pending Instance Ids that are known in this AWS State
+ * These are not categorized by region. It's one list of strings.
+ */
+AwsState.prototype.listPendingInstanceIds = function() {
+  var allIds = [];
+  var that = this;
+
+  this.regions().forEach(function(region) {
+    that.typesForRegion(region).forEach(function(workerType) {
+      var ids = that.__state[region][workerType].pending.map(function(x) {
+        return x.InstanceId;
+      });
+      Array.prototype.push.apply(allIds, ids);
+    });
+  });
+
+  return allIds;
+};
+
+
+/**
+ * Return a list of all Spot Request Ids that are known in this AWS State
+ * These are not categorized by region or by instance type. It's one
+ * list of strings.
+ */
+AwsState.prototype.listSpotRequestIds = function() {
+  var allIds = [];
+  var that = this;
+
+  this.regions().forEach(function(region) {
+    that.typesForRegion(region).forEach(function(workerType) {
+      var ids = that.__state[region][workerType].spotReq.map(function(x) {
+        return x.SpotInstanceRequestId;
+      });
+      Array.prototype.push.apply(allIds, ids);
+    });
+  });
+
+  return allIds;
+};
+
+
+/**
+ * Count the capacity of this workerType that are in the states specified
+ * by `states`.  Doing this uses the Capcity key from the workerType's
+ * types dictionary.  Remember that capacity is the number of tasks
+ * that this instance/request will be able to service.
+ * If specified, `extraSpotRequests` is a dictionary which contains a region
+ * and worker type categorized list of outstanding spot requests
+ */
+AwsState.prototype.capacityForType = function(workerType, extraSpotRequests, states) {
```

Comment:
Again, a good note, if you tracked all aws state in `AwsState` and used an `update` method, you wouldn't have to inject `extraSpotRequests`. And the data structure for this state object would be strictly limited the `AwsState` object. Granted one would have to notify the `AwsState` object when submitting a spot request. Or do the spot request using a method on `AwsState` like `AwsState.requestSpotInstance(launchConfiguration)`. Then create the launchConfig, etc. somewhere else..